### ISSUE TITLE
A quick thank-you & clarification on `MetricEmbedding`

### ISSUE THREAD

**Issue by tzslg** [wangshu31/MetricGrids#2]
Hi, thanks for sharing this interesting code!

I'm curious about the `MertricEmbedding` module — especially the `forward()` function. I noticed it applies three specific transformations:

* normalization to `[0, 1]`
* sine-based shift to `[-1, 0]`
* arcsin-based shift to `[-1, -0.5]`

Could you briefly explain the motivation or intuition behind these specific embeddings?
I appreciate the originality of this design — very cool idea!

Thanks again for your great work!

**Comment by wangshu31** [comment]
Hi, thank you very much for your interest in our work.

These transformations originate from Section 3.3 of the paper. As described, elementary metric grids of different orders exhibit varying degrees of sparsity. To handle this, we initially adopted multiple `tcnn.Encoding` objects with progressively smaller hash table lengths to store these features separately. However, I found this approach made it more difficult for the model to converge.

Instead, by transforming the coordinates of different metric grids into different ranges, we can partition a single hash table into segments of varying lengths — achieving a similar effect. This also helps keep the overall parameter count more manageable. It should be acknowledged that this alternative approach has not yet been thoroughly analyzed.

Thanks again for your insightful question!

**Comment by tzslg** [comment]
Hi, thank you again for your detailed explanation — I really appreciate your time and thoughtful insights!

I found Figure 2 in the paper particularly helpful for understanding the role of different metric spaces. One follow-up question: how are the different metric formulations (such as the generalized distance $|x - x_i|_p^p$) reflected in the implementation? Is this behavior approximated or absorbed by the choice of nonlinear transformations?

Also, I’ve been studying the MertricEmbedding module and have written a simplified version based on the original code:

```
class MertricEmbedding(torch.nn.Module):
    def __init__(self):
        super(MertricEmbedding, self).__init__()
        self.sin_1 = torch.sin(torch.tensor(1.0))
    def forward(self, x):
        out = [
            (x + 1.0) / 2.0,
            (torch.sin(x) - self.sin_1) / (2.0 * self.sin_1),
            (torch.arcsin(x) - (3.0 * torch.pi / 2.0)) / (2.0 * torch.pi)
        ]
        return torch.stack(out, dim=0)

```
Could you kindly help explain the reasoning behind this design? What role do these specific transformations play, and how do they approximate or emulate the behavior of different metric spaces described in the paper?

Thanks again for the inspiring work — I'm learning a lot from your implementation!

**Comment by wangshu31** [comment]
Hi, as you mentioned, different metric formulations are approximated via nonlinear transformations of coordinates. Faithfully implementing nonlinear metric formulas in PyTorch would inevitably increase inference time, which is particularly unfavorable for grid-based neural representations that already rely on a large number of parameters to achieve fast convergence.

As discussed in Section 4.1 of the paper, nonlinear metrics like $|x - x_i|_p^p$ are approximated by applying nonlinear transformations such as $x^p$ to the coordinates. In practice, we use nonlinear functions like $\sin(\cdot)$ and $\arcsin(\cdot)$. These provide stronger nonlinearity, as they can be further expanded into polynomial functions. While this choice is not unique, the nonlinearity it introduces is consistent.

Thank you again for your interest in the project!

**Comment by tzslg** [comment]
Hi, thank you for your clear and detailed explanation!
